add scan over layers section (#3195)
google/flax: docs/guides/haiku_migration_guide.rst

The only notable change with respect to the examples in the previous sections is that
this time around we used ``hk.without_apply_rng`` in Haiku so we didn't have to
pass the ``rng`` argument as ``None`` to the ``apply`` method.

> Scan Over Layers

Reviewer: ```suggestion
Scan over layers
```

We can implement other style guide-related changes to the rest of the doc in another PR.

----------------

> One very important application of ``scan`` is apply a sequence of layers iteratively

Reviewer: There's `jax.lax.scan` and `jax.linen.scan`, so let's use full API names. Also a small grammar fix.

```suggestion
One very important application of ``flax.linen.scan`` is to apply a sequence of layers iteratively
```

over an input, passing the output of each layer as the input to the next layer. This

> very useful to reduce compilation time for big models. As an example we will create

Reviewer: ```suggestion
is very useful to reduce compilation time for big models. As an example we will create
```

a simple ``Block`` Module, and then use it inside an ``MLP`` Module that will apply
the ``Block`` Module ``num_layers`` times.

In Haiku, we define the ``Block`` Module as usual, and then inside ``MLP`` we will
use ``hk.experimental.layer_stack`` over a ``stack_block`` function to create a stack
of ``Block`` Modules. In Flax, the definition of ``Block`` is a little different,
``__call__`` will accept and return a second dummy input/output that in both cases will

> be ``None``. In ``MLP``, we will use ``nn.scan`` as in the previous example, but

Reviewer: ```suggestion
be ``None``. In ``MLP``, we will use ``flax.linen.scan`` as in the previous example. But
```

by setting ``split_rngs={'params': True}`` and ``variable_axes={'params': 0}``

> we are telling ``nn.scan`` create different parameters for each step and slice the

Reviewer: ```suggestion
we are instructing ``flax.linen.scan`` to create different parameters for each step and slice the
```

      x = jax.nn.relu(x)
      return x

  class MLP(hk.Module):
    def __init__(self, features: int, num_layers: int, name=None):
        super().__init__(name=name)
        self.features = features
        self.num_layers = num_layers

    def __call__(self, x, training: bool):
     @hk.experimental.layer_stack(self.num_layers)
      def stack_block(x):
        return Block(self.features)(x, training)

      stack = hk.experimental.layer_stack(self.num_layers)
      return stack_block(x)

  ---

  class Block(nn.Module):
    features: int
    training: bool

    @nn.compact
    def __call__(self, x, _):
      x = nn.Dense(self.features)(x)
      x = nn.Dropout(0.5)(x, deterministic=not self.training)
      x = jax.nn.relu(x)
      return x, None

  class MLP(nn.Module):
    features: int
    num_layers: int

    @nn.compact
    def __call__(self, x, training: bool):
      ScanBlock = nn.scan(
        Block, variable_axes={'params': 0}, split_rngs={'params': True},
        length=self.num_layers)

      y, _ = ScanBlock(self.features, training)(x, None)
      return y

Notice how in Flax we pass ``None`` as the second argument to ``ScanBlock`` and ignore

> its second output, these represent the inputs/outputs per-step but they are ``None``

Reviewer: ```suggestion
its second output. These represent the inputs/outputs per-step but they are ``None``
```
